### PR TITLE
Fix gcc compile error

### PR DESCRIPTION
**Pull request by glingyan**
Fix Ubuntu gcc compile error &  mv mkl_memory.h to src\operator\mkl

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -61,6 +61,7 @@ endif
 ifeq ($(USE_MKL2017), 1)
 	CFLAGS += -DMXNET_USE_MKL2017=1
 	CFLAGS += -DUSE_MKL=1
+	CFLAGS += -I$(ROOTDIR)/src/operator/mkl/
```
Comment (piiswrong) [inline Makefile]: this should be in experimental right?

Comment (glingyan) [inline Makefile]: no , mkl_memory.h is need for all configuration

**Comment by piiswrong** [comment]
mkl batch norm is producing wrong results (tests/python/gpu/test_operator_gpu.py) Please fix asap

**Comment by glingyan** [comment]
Anything wrong , I do not have problem, below is my result

mxnet$ nosetests --verbose tests/python/gpu/test_operator_gpu.py
test_operator_gpu.test_batchnorm_with_type ... ok
test_operator_gpu.test_convolution_with_type ... MKL Build:20160921
ok
test_operator_gpu.test_deconvolution_with_type ... ok
test_operator_gpu.test_elementwise_sum ... ok
test_operator_gpu.test_upsampling_with_type ... ok
test_operator_gpu.test_concat_with_type ... ok
test_operator_gpu.test_elementwisesum_with_type ... ok
test_operator_gpu.test_reshape_with_type ... ok
test_operator_gpu.test_blockgrad_with_type ... ok
test_operator_gpu.test_concat ... ok
test_operator_gpu.test_swapaxis_with_type ... ok
test_operator_gpu.test_fullyconnected_with_type ... ok
test_operator_gpu.test_activation_with_type ... ok
test_operator_gpu.test_embedding_with_type ... ok
test_operator_gpu.test_slice_channel ... ok
test_operator_gpu.test_regression ... ok
test_operator_gpu.test_softmax ... ok
test_operator_gpu.test_python_op ... ok
test_operator_gpu.test_swapaxes ... ok
test_operator_gpu.test_scalarop ... ok
test_operator_gpu.test_scalar_pow ... ok
test_operator_gpu.test_symbol_pow ... ok
test_operator_gpu.test_pow_fn ... ok
test_operator_gpu.test_embedding ... ok
test_operator_gpu.test_binary_op_duplicate_input ... ok
test_operator_gpu.test_sign ... ok
test_operator_gpu.test_round_ceil_floor ... ok
test_operator_gpu.test_rsqrt_cos_sin ... ok
test_operator_gpu.test_maximum_minimum ... ok
test_operator_gpu.test_maximum_minimum_scalar ... ok
test_operator_gpu.test_abs ... ok
test_operator_gpu.test_deconvolution ... ok
test_operator_gpu.test_nearest_upsampling ... ok
test_operator_gpu.test_batchnorm_training ... ok
test_operator_gpu.test_convolution_grouping ... ok
test_operator_gpu.test_broadcast_binary_op ... ok
test_operator_gpu.test_run_convolution_dilated_impulse_response ... ok
test_operator_gpu.test_convolution_dilated_impulse_response ... ok
test_operator_gpu.test_reshape ... [23:45:22] src/operator/./reshape-inl.h:270: Using target_shape will be deprecated.
ok
test_operator_gpu.test_reduce ... ok
test_operator_gpu.test_broadcast ... ok
test_operator_gpu.test_transpose ... ok
test_operator_gpu.test_expand_dims ... ok
test_operator_gpu.test_crop ... ok
test_operator_gpu.test_slice_axis ... ok
test_operator_gpu.test_flip ... ok
test_operator_gpu.test_stn ... ok
test_operator_gpu.test_dot ... ok
test_operator_gpu.test_batch_dot ... ok
test_operator_gpu.test_correlation ... ok
test_operator_gpu.test_roipooling ... ok
test_operator_gpu.test_pad ... ok
test_operator_gpu.test_instance_normalization ... ok

----------------------------------------------------------------------
Ran 53 tests in 53.830s

OK

**Comment by glingyan** [comment]
This is my configuration
diff --git a/make/config.mk b/make/config.mk
index 8c4dfd4..2144110 100644
--- a/make/config.mk
+++ b/make/config.mk
@@ -38,15 +38,15 @@ ADD_CFLAGS =
 #---------------------------------------------
 
 # whether use CUDA during compile
-USE_CUDA = 0
+USE_CUDA = 1
 
 # add the path to CUDA library to link and compile flag
 # if you have already add them to environment variable, leave it as NONE
-# USE_CUDA_PATH = /usr/local/cuda
-USE_CUDA_PATH = NONE
+USE_CUDA_PATH = /usr/local/cuda
+#USE_CUDA_PATH = NONE
 
 # whether use CuDNN R3 library
-USE_CUDNN = 0
+USE_CUDNN = 1
 
 # whether use cuda runtime compiling for writing kernels in native language (i.e. Python)
 USE_NVRTC = 0
@@ -60,10 +60,10 @@ USE_OPENCV = 1
 USE_OPENMP = 1
 
 # whether use MKL2017 library
-USE_MKL2017 = 0
+USE_MKL2017 = 1
 
 # whether use MKL2017 experimental feature for high performance
-USE_MKL2017_EXPERIMENTAL = 0
+USE_MKL2017_EXPERIMENTAL = 1
 
 # choose the version of blas you want to use
 # can be: mkl, blas, atlas, openblas
@@ -72,7 +72,7 @@ UNAME_S := $(shell uname -s)
 ifeq ($(UNAME_S), Darwin)
 USE_BLAS = apple
 else
-USE_BLAS = atlas
+USE_BLAS = mkl
 endif
 
 # add path to intel library, you may need it for MKL, if you did not add the path

**Comment by piiswrong** [comment]
try
`python tests/python/gpu/test_operator_gpu.py`

nosetests wasn't working correctly. I just fixed it

**Comment by piiswrong** [comment]
any updates?

**Comment by glingyan** [comment]
HI Eric:

   Root cause found
   MXNet’s test case require to support moving mean &var as the same as GPU for both fix_gamma & not fix gamma case
   MKL team will provide the feature these weeks, and I will enable it as soon as possible.
   Below is the function to check consistency
def check_consistency(sym, ctx_list, scale=1.0, grad_req='write'):
   tol = {np.dtype(np.float16): 1e-1,
           np.dtype(np.float32): 1e-3,
           np.dtype(np.float64): 1e-5,
           np.dtype(np.uint8): 0,
           np.dtype(np.int32): 0}
    assert len(ctx_list) > 1
    exe_list = [sym.simple_bind(grad_req=grad_req, **ctx) for ctx in ctx_list]
    for exe in exe_list:
        assert len(exe.outputs) == 1
        assert len(exe.arg_arrays) == len(exe_list[0].arg_arrays)
        assert len(exe.grad_arrays) == len(exe_list[0].grad_arrays)

    init = [np.random.normal(size=arr.shape, scale=scale) for arr in exe_list[0].arg_arrays]
    if sym.name == 'embedding':
        init[0] = np.random.randint(low=0, high=10, size=exe_list[0].arg_arrays[0].shape)

    for exe in exe_list:
        for arr, iarr in zip(exe.arg_arrays, init):
            arr[:] = iarr.astype(arr.dtype)

    outputs_0 = [exe.outputs[0].asnumpy() for exe in exe_list]
    grads_0 = [[grad.asnumpy() if grad is not None else np.zeros(1) for grad in exe.grad_arrays] for exe in exe_list]
    # forward  //Not problem , since moving mean is 0 at the begining
    for exe in exe_list:
        exe.forward(is_train=True)
        exe.backward(exe.outputs[0])

    outputs = [exe.outputs[0].asnumpy() for exe in exe_list]
    # lazy solution handling None grad
    grads = [[grad.asnumpy() if grad is not None else np.zeros(1) for grad in exe.grad_arrays] for exe in exe_list]
    dtypes = [arr.dtype for arr in outputs]
    max_idx = np.argmax(dtypes)


    //arg stand for BatchNormOpInputs {kData, kGamma, kBeta};
    arg_arrays_1_0 = [exe.arg_arrays[0].asnumpy() for exe in exe_list]
    arg_arrays_1_1 = [exe.arg_arrays[1].asnumpy() for exe in exe_list]
arg_arrays_1_2 = [exe.arg_arrays[2].asnumpy() for exe in exe_list]
//aux stand for BatchNormOpAuxiliary {kMovingMean, kMovingVar};
    aux_arrays_1_0 = [exe.aux_arrays[0].asnumpy() for exe in exe_list]
    aux_arrays_1_1 = [exe.aux_arrays[1].asnumpy() for exe in exe_list]
[cid:image001.png@01D23E52.9F809450]
    #forward predict
    for exe in exe_list:  //Have problem here
        exe.forward(is_train=False)
